crates/sui-framework/sources/collectible/kiosk.move

```diff
@@ -0,0 +1,474 @@
+// Copyright (c) Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+/// Kiosk is a primitive and an open, zero-fee trading platform
+/// for assets with high degree of customization over transfer
+/// policies.
+///
+/// The system has 3 main audiences:
+///
+/// 1. Creators: for a type to be tradable in the Kiosk ecosystem,
+/// creator (publisher) of the type needs to issue a `TransferPolicyCap`
+/// which gives them a power to enforce any constraint on trades by
+/// either using one of the pre-built primitives (see `sui::royalty`)
+/// or by implementing a custom policy. The latter requires additional
+/// support for discoverability in the ecosystem and should be performed
+/// within the scope of an Application or some platform.
+///
+/// - A type can not be traded in the Kiosk unless there's a policy for it.
+/// - 0-royalty policy is just as easy as "freezing" the `AllowTransferCap`
+///   making it available for everyone to authorize deals "for free"
+///
+/// 2. Traders: anyone can create a Kiosk and depending on whether it's
+/// a shared object or some shared-wrapper the owner can trade any type
+/// that has issued `TransferPolicyCap` in a Kiosk. To do so, they need
+/// to make an offer, and any party can purchase the item for the amount of
+/// SUI set in the offer. The responsibility to follow the transfer policy
+/// set by the creator of the `T` is on the buyer.
+///
+/// 3. Marketplaces: marketplaces can either watch for the offers made in
+/// personal Kiosks or even integrate the Kiosk primitive and build on top
+/// of it. In the custom logic scenario, the `TransferPolicyCap` can also
+/// be used to implement application-specific transfer rules.
+///
+module sui::kiosk {
+    use std::option::{Self, Option};
+    use sui::object::{Self, UID, ID};
+    use sui::dynamic_object_field as dof;
+    use sui::dynamic_field as df;
+    use sui::package::{Self, Publisher};
+    use sui::tx_context::{TxContext, sender};
+    use sui::balance::{Self, Balance};
+    use sui::coin::{Self, Coin};
+    use sui::sui::SUI;
+    use sui::event;
+
+    // Collectible is a special case to avoid storing `Publisher`.
+    friend sui::collectible;
+
+    /// For when trying to withdraw profits as owner and owner is not set.
+    const EOwnerNotSet: u64 = 0;
+    /// For when trying to withdraw profits and sender is not owner.
+    const ENotOwner: u64 = 1;
+    /// For when Coin paid does not match the offer price.
+    const EIncorrectAmount: u64 = 2;
+    /// For when incorrect arguments passed into `switch_mode` function.
+    const EIncorrectArgument: u64 = 3;
+    /// For when Transfer is accepted by a wrong Kiosk.
+    const EWrongTarget: u64 = 4;
+    /// For when trying to withdraw higher amount than stored.
+    const ENotEnough: u64 = 5;
+    /// For when trying to close a Kiosk and it has items in it.
+    const ENotEmpty: u64 = 6;
```

Comment:
nit: drop "For when"? IIRC, these comments can get used by the source map to symbolicate aborts, so it's good when they can serve as both the docs for the constant and the error message.